snakemake/script.py

```diff
@@ -381,42 +410,112 @@ def generate_preamble(
-            resources,
-            log,
-            config,
-            rulename,
-            bench_iteration,
-            os.path.dirname(wrapper_path),
-        )
-        snakemake = pickle.dumps(snakemake)
-        # Obtain search path for current snakemake module.
-        # The module is needed for unpickling in the script.
-        # We append it at the end (as a fallback).
+
+        # Obtain search path for any mounted containers
+        # This will also ensure recipes that require snakemake have it on path
         searchpath = SNAKEMAKE_SEARCHPATH
         if container_img is not None:
             searchpath = singularity.SNAKEMAKE_MOUNTPOINT
         searchpath = repr(searchpath)
+
         # For local scripts, add their location to the path in case they use path-based imports
         if path.startswith("file://"):
             searchpath += ", " + repr(os.path.dirname(path[7:]))
 
+        # The wrapper path should be added too
+        wrapper_path = path[7:] if path.startswith("file://") else path
+
         return textwrap.dedent(
             """
         ######## snakemake preamble start (automatically inserted, do not edit) ########
-        import sys; sys.path.extend([{searchpath}]); import pickle; snakemake = pickle.loads({snakemake}); from snakemake.logging import logger; logger.printshellcmds = {printshellcmds}; {preamble_addendum}
+        import sys, os; sys.path.extend([{searchpath}]); sys.path.insert(0, {sourcedir});from collections import namedtuple
+        class Snakemake:
+            def __init__(self, input, output, params, wildcards, threads, log, resources, config, rule, bench_iteration, scriptdir, source=None):
+                self.set_listvar("input", input)
+                self.set_listvar("params", params)
+                self.output = output
+                self.wildcards = wildcards
+                self.threads = threads
+                self.log = log[0]
+                self.resources = resources
+                self.config = config
+                self.rule = rule
+                self.bench_iteration = bench_iteration
+                self.scriptdir = scriptdir
+                if source is not None:
+                    self.source = source
+
+            def set_listvar(self, name, items):
+                class Item:
+                    def __init__(self, entries):
+                        self.__dict__.update(entries)
+                    def get(self, key, default=None):
+                        if hasattr(self, key):
+                            return getattr(self, key)
+                        return default
+
+                structure = Item(items)
+                setattr(self, name, structure)
+ 
+            def source(self):
+                wd = os.getcwd()
+                os.chdir(self.scriptdir)
+                sys.path.insert(self.scriptdir)
+                os.chdir(wd)
+                
+            def log_fmt_shell(self, stdout=True, stderr=True, append=False):
```

Comment:
This had a format / weird error I couldn't figure out when it was represented as it was before (with the Snakemake object). When I wrote it out like this, it worked ok.